src/Proxy.NET/Program.cs

```diff
@@ -0,0 +1,177 @@
+        dataSourceBuilder.MapEnum<ModelType>();
+
+        if (string.IsNullOrEmpty(dataSourceBuilder.ConnectionStringBuilder.Password))
+        {
+            // If no password was provided, we're probably going to be using a managed identity to authenticate.
+            // Define the time period the token lives for, and how long to wait to retry if the token fails.
+            int refreshPeriod = builder.Configuration.GetValue("Azure:TokenRefreshSeconds", 14400);
+            int failureRefreshPeriod = builder.Configuration.GetValue("Azure:TokenFailureRefreshSeconds", 10);
+
+            // Use the Azure SDK to get a token from the managed identity and provide it to Npgsql using the
+            // PeriodicPasswordProvider to refresh it on schedule.
+            dataSourceBuilder.UsePeriodicPasswordProvider(
+                async (_, ct) =>
+                {
+                    var credential = new DefaultAzureCredential();
+                    var ctx = new TokenRequestContext(["https://ossrdbms-aad.database.windows.net/.default"]);
+                    var tokenResponse = await credential.GetTokenAsync(ctx, ct);
+                    return tokenResponse.Token;
+                },
+                TimeSpan.FromSeconds(refreshPeriod),
+                TimeSpan.FromSeconds(failureRefreshPeriod)
+            );
+        }
+
+        NpgsqlDataSource dataSource = dataSourceBuilder.Build();
+
+        builder.Services.AddDbContext<AoaiProxyContext>(
+            (option) =>
+            {
+                option.UseNpgsql(
+                    dataSource,
+                    // https://stackoverflow.com/questions/70423137/how-to-gracefully-handle-a-postgres-restart-in-npgsql
+                    (options) =>
+                    {
+                        options.EnableRetryOnFailure(4, TimeSpan.FromSeconds(30), new string[] { "57P01" });
+                    }
+                );
+            }
+        );
+
+        // Add services to the container.
+        builder.Services.AddAuthorization();
+        builder.Services.AddMemoryCache();
+
+        // Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
+        builder.Services.AddEndpointsApiExplorer();
+        builder.Services.AddSwaggerGen();
+
+        // https://learn.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests
+        builder.Services.AddHttpClient<IProxyService, ProxyService>();
+
+        builder.Services.AddScoped<ICatalogService, CatalogService>();
+        builder.Services.AddScoped<IAuthorizeService, AuthorizeService>();
+        builder.Services.AddScoped<IProxyService, ProxyService>();
+        builder.Services.AddScoped<IMetricService, MetricService>();
+        builder.Services.AddScoped<IAttendeeService, AttendeeService>();
+        builder.Services.AddScoped<IEventService, EventService>();
```

Comment:
A good idea would be to roll this out into a separate extension method, like you did with the endpoint registration.

Reply:
fixed